File: Excercise 6/exercise.cpp
#include <iostream>
#include <unordered_map>

using namespace std;

int majorityElement(int arr[], int size) {
    // Solving using unordered_map.
    unordered_map<int, int> m;

    // Iterate through each element and add it to the unordered_map.
    for (int i = 0; i < size; i++) {
        m[arr[i]]++;
    }

    // Variable to store the resut.
    int res;

    // Iterate through unordered_map.
    for (auto x : m) {
        // If the occurence of a value is greater than size / 2, then that is the majority element.
        if (x.second > size/2) {
            res = x.first;
        }
    }

    // Return the resut.
    return res;
}

int main() {
    cout << "Case 1" << endl;
    int arr[] = {3, 2, 3};
    int size = sizeof(arr)/sizeof(arr[0]);
    cout << majorityElement(arr, size) << endl;
    // Expected : 3
    // Result   : 3

    cout << "Case 2" << endl;
    int arr2[] = {2,2,1,1,1,2,2};
    size = sizeof(arr2)/sizeof(arr2[0]);
    cout << majorityElement(arr2, size) << endl;
    // Expected : 2
    // Result   : 2

    return 0;
}

File: Excercise 9/exercise.java
public class App {
    public static void main(String[] args) {
        int[] arr = {1,2,3,10};
        System.out.println("Case 1");
        System.out.println(smallestPostiveInt(arr));
        // Expected : 7
        // Result   : 7

        System.out.println("Case 2");
        int[] arr2 = {1,2,3,4,13};
        System.out.println(smallestPostiveInt(arr2));
        // Expected : 11
        // Result   : 11

        System.out.println("Case 3");
        int[] arr3 = {1};
        System.out.println(smallestPostiveInt(arr3));
        // Expected : 2
        // Result   : 2

        System.out.println("Case 4");
        int[] arr4 = {};
        System.out.println(smallestPostiveInt(arr4));
        // Expected : 1
        // Result   : 1
    }

    // Method to find the smallest postive int that is not a sum of any subset of the array.
    public static int smallestPostiveInt(int[] arr) {
        // First we initialize our result to be 1 because that is the smallest number wer can get. 
        int res = 1;

        // Since array is already sorted, just run thorugh the loop.
        for (int i = 0; i < arr.length; i++) {
            // If the element is greate than the result variable, then we have the smallest number.
            if (arr[i] > res) {
                // Return the smallest number.
                return res;
            
            // If not, then we add the result variable to the current element of the array.
            // That would be the next smallest number.
            } else {
                res = arr[i] + res;
            }
        }

        // Return the result variable.
        return res;
    }
}